File: src/main/java/leetCode/offer/Offer22.java
package leetCode.offer;

/**
 * @author jingxinwu
 * @date 2021-11-21 6:07 下午
 */
public class Offer22 {

    public ListNode getKthFromEnd(ListNode head, int k) {
        int n = 0;
        ListNode node;
        for (node = head; node != null; node = node.next) {
            n++;
        }
        for (node = head; n > k; n--) {
            node = node.next;
        }
        return node;
    }
}

File: src/main/go/leetcode/lists/offerII/Offer22.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/13 16:10
 */

func detectCycle(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			fast = head

			for slow != fast {
				slow = slow.Next
				fast = fast.Next
			}
			return slow
		}
	}
	return nil
}
